### PR TITLE
🌱  Setup PubSub framework code

### PR DESCRIPTION
**Pull request by azeemshaikh38**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Adds framework/skeleton code for setting up PubSub batch job as discussed in https://github.com/ossf/scorecard/issues/318#issuecomment-832348904. This PR also includes code for Master server which publishes data to the repo_request topic.

* **What is the current behavior?** (You can also link to an open issue here)
N/A

* **What is the new behavior (if this is a feature change)?**
No new behaviour added yet in this PR.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

### CODE REVIEW

**Comment by naveensrinivasan** [comment]
@azeemshaikh38  Some documentation on a large PR would help with the review.  The approach and your thought process. 

For example here is HTTP caching PR https://github.com/ossf/scorecard/pull/203 and some related docs https://github.com/ossf/scorecard#caching
https://github.com/ossf/scorecard/issues/80#issuecomment-782723182

I would suggest that some docs be included for review as well as others to contribute in the future.

**Comment by azeemshaikh38** [comment]
Re: large PR comment. Each individual piece didn't make sense on its own, so had to wait to send this out only after all parts were in order. 

Re: documentation. The PR is based on this comment - https://github.com/ossf/scorecard/issues/318#issuecomment-832348904. What this is PR trying to achieve is - given a list of repos, "master/main.go" will send out requests to a PubSub topic. To do this, it follows these steps:

1. Uses "data/reader.go" to read from the input file. "data/reader.go" abstracts away the format of the file itself and provides a simple HasNext(), GetNext() type interface.
2. Uses "data/request.proto" to create a request proto. Each request corresponds to a single JSON shard file in  https://github.com/ossf/scorecard/issues/318#issuecomment-832348904.
3. Uses "data/publisher.go" to publish to a given topic.
4. Uses "data/blob.go" to write some metadata once all requests are successfully sent. 

I have tried to be as generic as possible to avoid GoogleCloud specific logic. So as much as possible, logic is encapsulated inside gocloud.dev packages. The config/config.go serves a single place to update and use different cloud services if need be.

**Comment by naveensrinivasan** [comment]
> Re: large PR comment. Each individual piece didn't make sense on its own, so had to wait to send this out only after all parts were in order.
> 
> Re: documentation. The PR is based on this comment - [#318 (comment)](https://github.com/ossf/scorecard/issues/318#issuecomment-832348904). What this is PR trying to achieve is - given a list of repos, "master/main.go" will send out requests to a PubSub topic. To do this, it follows these steps:
> 
> 1. Uses "data/reader.go" to read from the input file. "data/reader.go" abstracts away the format of the file itself and provides a simple HasNext(), GetNext() type interface.
> 2. Uses "data/request.proto" to create a request proto. Each request corresponds to a single JSON shard file in  [#318 (comment)](https://github.com/ossf/scorecard/issues/318#issuecomment-832348904).
> 3. Uses "data/publisher.go" to publish to a given topic.
> 4. Uses "data/blob.go" to write some metadata once all requests are successfully sent.
> 
> I have tried to be as generic as possible to avoid GoogleCloud specific logic. So as much as possible, logic is encapsulated inside gocloud.dev packages. The config/config.go serves a single place to update and use different cloud services if need be.

Thanks this helps!

File: cron/data/blob.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ossf/scorecard/cron/config"
+	"gocloud.dev/blob"
+)
+
+const (
+	filePrefixFormat = "2006.01.02/150405/"
```
Comment (naveensrinivasan) [inline cron/data/blob.go:28]: Some documentation on the filePrefixformat will help.

Comment (naveensrinivasan) [reply cron/data/blob.go:28]: What is `150405` signify? It is not clear.

File: cron/data/blob.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		return fmt.Errorf("%w: %v", errorWriteClose, err)
+	}
+	return nil
+}
+
+func GetBlobFilename(filename string, datetime time.Time) string {
+	return datetime.Format(filePrefixFormat) + filename
+}
+
+func GetShardNumFilename(datetime time.Time) string {
```
Comment (naveensrinivasan) [inline cron/data/blob.go:61]: Docs on this function would be helpful. Probably how and the data is going to be sharded.

File: cron/data/publisher.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		return &ret, fmt.Errorf("%w: %v", errorOpenTopic, err)
+	}
+	return &publisherImpl{
+		ctx:   ctx,
+		topic: topic,
+	}, nil
+}
+
+type topicInterface interface {
```
Comment (naveensrinivasan) [inline cron/data/publisher.go:53]: Instead of `topicInterface` could be just `publisher`? And the method be `publish`.
 https://golang.org/doc/effective_go#interface-names
 
In golang By convention, one-method interface names are nouns denoting the doer of an action.

File: cron/data/reader.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"io"
+
+	"github.com/jszwec/csvutil"
+	"github.com/ossf/scorecard/repos"
+)
+
+var errorCSVDecoder = errors.New("could not create CSV decoder")
+
+type Reader interface {
```
Comment (naveensrinivasan) [inline cron/data/reader.go:29]: > By convention, one-method interfaces are named by the method name plus an -er suffix or similar modification to construct an agent noun: Reader, Writer, Formatter, CloseNotifier etc.

https://golang.org/doc/effective_go#interface-names

Why not use the standard `io.Reader` interface? It is used across standard lib and lots of things are built on this convention. https://golang.org/pkg/io/


The iterator pattern used in other languages is not idiomatic go because other things aren't built around this way.

**Review by naveensrinivasan — CHANGES_REQUESTED** [review]
I like the basic structure. Expect few things that aren't idiomatic go and few nits.

File: cron/data/blob.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+var (
+	errorWrite      = errors.New("blobWriter.Write")
+	errorWriteClose = errors.New("blobWriter.Close")
+	errorOpenBucket = errors.New("blob.OpenBucket")
+)
+
+func WriteToBlobStore(ctx context.Context, bucketURL, filename string, data []byte) error {
+	bucket, err := blob.OpenBucket(ctx, bucketURL)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errorOpenBucket, err)
```
Comment (oliverchang) [inline cron/data/blob.go:40]: nit: I don't think these fmt.Errorf calls are correct. See examples in https://blog.golang.org/go1.13-errors#TOC_3.3.

You should only have 1 error wrapped by %w, and this should be the underlying `err` that you got back from the previous call.


i.e.

`fmt.Errorf("blob.OpenBucket failed: %w", err)`

File: cron/master/main.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Security Scorecard Authors
```
Comment (oliverchang) [inline cron/master/main.go:1]: nit: consider a name other than "master" for this. Maybe "controller" ?

File: cron/master/main.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+			continue
+		}
+		request.Repos = append(request.GetRepos(), repoURL.URL())
+		if len(request.GetRepos()) < config.ShardSize {
+			continue
+		}
+		if err := topicPublisher.Publish(&request); err != nil {
+			return shardNum, fmt.Errorf("%w: %v", errorPublishing, err)
+		}
+		_ = request.Repos[:0]
```
Comment (oliverchang) [inline cron/master/main.go:59]: what's this for?

Comment (oliverchang) [inline cron/master/main.go:59]: do you mean to increment shardnum here?

File: cron/master/main.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	t := time.Now()
+	reposFile, err := os.OpenFile(config.InputReposFile, os.O_RDONLY, 0o644)
+	if err != nil {
+		panic(err)
+	}
+	if reader, err := data.MakeReader(reposFile); err != nil {
+		panic(err)
+	} else if shardNum, err := PublishToRepoRequestTopic(ctx, reader, t); err != nil {
+		panic(err)
+	} else if err := data.WriteToBlobStore(ctx,
```
Comment (oliverchang) [inline cron/master/main.go:78]: the logic is actually kind of hard to follow here when written this way. this is probably more readable without the "else ifs".

**Review by oliverchang — COMMENTED** [review]
nice! some cursory comments. will take a closer look tomorrow.

File: cron/data/reader.go (hunk 4 above)
Comment (azeemshaikh38) [reply cron/data/reader.go:29]: I don't agree with iterator pattern not being a Go way. I see popular Go packages use this pattern, including Google APIs - https://github.com/googleapis/google-cloud-go/wiki/Iterator-Guidelines

However, changing the interface name to `Iterator` to better reflect its purpose better.

File: cron/data/blob.go (hunk 2 above)
Comment (azeemshaikh38) [reply cron/data/blob.go:63]: I want to avoid obvious comments in code. The fn does precisely as the name suggests, i.e - given a `time.Time` value, returns a filename `string` which in this case, is a pre-defined metadata file called `.shard_num`. It does not impact data sharding.

File: cron/config/config.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permisshandlerns and
+// limitathandlerns under the License.
+
+package config
+
+const (
+	ResultDataBucketURL string = "gs://ossf-scorecard-data"
+	RequestTopicURL            = "gcppubsub://projects/openssf/topics/scorecard-batch"
```
Comment (inferno-chromium) [inline cron/config/config.go:19]: openssf should be moved to its own var and then RequestTopicURL should embed it.

File: cron/data/publisher.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	"log"
+	"sync"
+	"sync/atomic"
+
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/gcppubsub" // Needed to link in GCP drivers.
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var errorPublish = errors.New("total errors when publishing")
```
Comment (inferno-chromium) [inline cron/data/publisher.go:30]: seems used at only one site, why not just put it there.

File: cron/data/iterator_test.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+	}{
+		{
+			name:     "Basic",
+			filename: "testdata/basic.csv",
+			outcomes: []outcome{
+				{
+					hasError: false,
+					repo: repos.RepoURL{
+						Host:  "github.com",
+						Owner: "18F",
```
Comment (inferno-chromium) [inline cron/data/iterator_test.go:44]: nit: should just use fake repo names instead of these weird repo names 18F/1N3, etc

File: cron/data/request.pb.go [hunk 12]
```diff
@@ -0,0 +1,180 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
```
Comment (inferno-chromium) [inline cron/data/request.pb.go:1]: Add a README.md file on how to generate this (command line used to generate this).

File: cron/data/publisher.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
+// Copyright 2021 Security Scorecard Authors
```
Comment (inferno-chromium) [inline cron/data/publisher.go:1]: publisher*.go seems weird to go in data/ dir, maybe something else?

**Review by inferno-chromium — DISMISSED** [review]

File: cron/master/main.go (hunk 8 above)
Comment (oliverchang) [reply cron/master/main.go]: yeah, that's just how Go is :)

File: cron/data/iterator.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	return reader.err != io.EOF
+}
+
+func (reader *csvReader) Next() (repos.RepoURL, error) {
+	if reader.err != nil {
+		return repos.RepoURL{}, reader.err
+	}
+	ret := repos.RepoURL{}
+	var err error
+	err = ret.Set(reader.next.Repo)
```
Comment (oliverchang) [inline cron/data/iterator.go:62]: nit: just `err := ` ?

**Review by oliverchang — APPROVED** [review]
nice!

**Review by naveensrinivasan — DISMISSED** [review]

File: cron/config/config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (inferno-chromium) [reply cron/config/config.go:19]: Ideally, project name and gcs bucket should go in a config.yaml and read from there. If some env var is set (e.g. someone's dockerfile who wants to run scorecard service on their project (although rare)), it can use it as an override, otherwise use the default from config.yaml. for tests, you can just set those env vars. this should avoid dependency injection. wdyt @naveensrinivasan @azeemshaikh38 @oliverchang

[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply cron/config/config.go:19]: I agree with that https://12factor.net/config. Thanks

[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply cron/config/config.go:19]: Without the `scorecard` word, the URL does not indicate that it is related to Scorecard project. How about `scorecard-requests` or `scorecard-batch-requests`?

File: cron/config/config.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+	InputReposFile             = "../projects.txt"
```
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply cron/config/config.go]: Done.

File: cron/controller/main.go [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+	request := data.ScorecardBatchRequest{
+		JobTime:  timestamppb.New(datetime),
+		ShardNum: &shardNum,
+	}
+	topicPublisher, err := data.CreatePublisher(ctx, config.RequestTopicURL)
+	if err != nil {
+		return shardNum, fmt.Errorf("error running CreatePublisher: %w", err)
+	}
+
+	for iter.HasNext() {
```
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply cron/controller/main.go:45]: Done.

File: cron/data/iterator.go (hunk 14 above)
Comment (azeemshaikh38) [reply cron/data/iterator.go:62]: Go complains with `no new variables on left side of :=` if I use `err := `. Maybe I'm missing some Go syntax here?

File: cron/data/iterator_test.go (hunk 11 above)
Comment (azeemshaikh38) [reply cron/data/iterator_test.go]: Done.

File: cron/data/publisher.go (hunk 13 above)
Comment (azeemshaikh38) [reply cron/data/publisher.go]: Done. Moved to pubsub/

File: cron/data/publisher.go (hunk 10 above)
Comment (azeemshaikh38) [reply cron/data/publisher.go]: This is to avoid the linter from complaining - dynamic errors are not allowed.

File: cron/config/config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (inferno-chromium) [reply cron/config/config.go:19]: sure, scorecard-batch-requests seems fine.

**Comment by naveensrinivasan** [comment]
There is a binary file in this commit https://github.com/ossf/scorecard/blob/0f01b1501a5f4dad1ef50a5f79edba08e0a82753/cron/cron which could be an accidental commit.

**Review by inferno-chromium — APPROVED** [review]
putting approval back as a setting change had reset it.

**Comment by azeemshaikh38** [comment]
> There is a binary file in this commit https://github.com/ossf/scorecard/blob/0f01b1501a5f4dad1ef50a5f79edba08e0a82753/cron/cron which could be an accidental commit.

Deleted.

File: cron/config/config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply cron/config/config.go]: I have added a TODO for this. Will have a follow up PR addressing this to avoid adding more logic in this PR.

Comment (azeemshaikh38) [reply cron/config/config.go]: Moving these vars to config.yaml in a follow up PR. Will address this there.

File: cron/data/request.pb.go (hunk 12 above)
Comment (azeemshaikh38) [reply cron/data/request.pb.go:15]: Done.
